hdlparse/vhdl_parser.py

```diff
@@ -6,748 +6,829 @@
 
-    def __init__(self, name, package, parameters, return_type=None, desc=None):
-        VhdlObject.__init__(self, name, desc)
-        self.kind = 'function'
-        self.package = package
-        self.parameters = parameters
-        self.return_type = return_type
+  else: # procedure
+    proto = 'procedure {}({});'.format(vo.name, plist)
 
-    def __repr__(self):
-        return "VhdlFunction('{}')".format(self.name)
+  return proto
 
+def subprogram_signature(vo, fullname=None):
+  '''Generate a signature string
 
-class VhdlProcedure(VhdlObject):
-    """Procedure declaration
+  Args:
+    vo (VhdlFunction, VhdlProcedure): Subprogram object
+  Returns:
+    Signature string.
+  '''
 
-    Args:
-      name (str): Name of the procedure
-      package (str): Package containing the procedure
-      parameters (list of VhdlParameter): Parameters to the procedure
-      desc (str, optional): Description from object metacomments
-    """
+  if fullname is None:
+    fullname = vo.name
 
-    def __init__(self, name, package, parameters, desc=None):
-        VhdlObject.__init__(self, name, desc)
-        self.kind = 'procedure'
-        self.package = package
-        self.parameters = parameters
+  if isinstance(vo, VhdlFunction):
+    plist = ','.join(p.data_type for p in vo.parameters)
+    sig = '{}[{} return {}]'.format(fullname, plist, vo.return_type)
+  else: # procedure
+    plist = ','.join(p.data_type for p in vo.parameters)
+    sig = '{}[{}]'.format(fullname, plist)
 
-    def __repr__(self):
-        return "VhdlProcedure('{}')".format(self.name)
+  return sig
 
 
-class VhdlComponent(VhdlObject):
-    """Component declaration
+def is_vhdl(fname):
+  '''Identify file as VHDL by its extension
 
-    Args:
-      name (str): Name of the component
-      package (str): Package containing the component
-      ports (list of VhdlParameter): Port parameters to the component
-      generics (list of VhdlParameter): Generic parameters to the component
-      sections (list of str): Metacomment sections
-      desc (str, optional): Description from object metacomments
-    """
-
-    def __init__(self, name, package, ports, generics=None, sections=None, desc=None):
-        VhdlObject.__init__(self, name, desc)
-        self.kind = 'component'
-        self.package = package
-        self.generics = generics if generics is not None else []
-        self.ports = ports
-        self.sections = sections if sections is not None else {}
-
-    def __repr__(self):
-        return "VhdlComponent('{}')".format(self.name)
-
-    def dump(self):
-        print('VHDL component: {}'.format(self.name))
-        for p in self.ports:
-            print('\t{} ({}), {} ({})'.format(p.name, type(p.name), p.data_type, type(p.data_type)))
+  Args:
+    fname (str): File name to check
+  Returns:
+    True when file has a VHDL extension.
+  '''
+  return os.path.splitext(fname)[1].lower() in ('.vhdl', '.vhd')
```

Comment:
`pathlib` from Python (introduced in Python 3.4) should be used for such operations.

Reply:
Forwarded to #5.